lib/core/src/Cardano/Wallet/DB/Store/Wallets/Store.hs

```diff
@@ -17,109 +19,159 @@ Implementation of a store for 'TxWalletsHistory'
+    ( EntityField (..), LocalTxSubmission, TxMeta )
 import Cardano.Wallet.DB.Store.Meta.Model
-    ( DeltaTxMetaHistory (Manipulate), TxMetaHistory, mkTxMetaHistory )
+    ( TxMetaHistory (TxMetaHistory), mkTxMetaHistory )
 import Cardano.Wallet.DB.Store.Meta.Store
     ( mkStoreMetaTransactions )
+import Cardano.Wallet.DB.Store.Submissions.Model
+    ( TxLocalSubmissionHistory (TxLocalSubmissionHistory) )
+import Cardano.Wallet.DB.Store.Submissions.Store
+    ( mkStoreSubmissions )
 import Cardano.Wallet.DB.Store.Transactions.Model
     ( DeltaTxHistory (..), TxHistoryF (..), mkTxHistory )
 import Cardano.Wallet.DB.Store.Transactions.Store
     ( mkStoreTransactions )
 import Cardano.Wallet.DB.Store.Wallets.Model
-    ( DeltaTxWalletsHistory (..), walletsLinkedTransactions )
+    ( DeltaTxWalletsHistory (..)
+    , DeltaWalletsMetaWithSubmissions (..)
+    , walletsLinkedTransactions
+    )
 import Control.Applicative
     ( liftA2 )
-import Control.Exception
-    ( SomeException )
 import Control.Monad
     ( forM, forM_ )
 import Control.Monad.Except
-    ( ExceptT (ExceptT), runExceptT )
+    ( ExceptT (ExceptT), lift, runExceptT )
 import Data.DBVar
     ( Store (..) )
+import Data.Delta
+    ( apply )
 import Data.DeltaMap
     ( DeltaMap (..) )
+import Data.Functor
+    ( ($>) )
 import Data.Generics.Internal.VL
     ( view )
 import Data.List
     ( nub )
+import Data.Map.Strict
+    ( Map )
 import Database.Persist.Sql
     ( SqlPersistT, deleteWhere, entityVal, selectList, (==.) )
 
 import qualified Cardano.Wallet.DB.Store.Meta.Model as TxMetaStore
 import qualified Cardano.Wallet.DB.Store.Transactions.Model as TxStore
 import qualified Cardano.Wallet.Primitive.Types as W
 import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
 
--- | Store for a map of 'DeltaTxMetaHistory' of multiple different wallets.
-mkStoreWalletsMeta :: Store
+
+mkStoreWalletMetaWithSubmissions :: W.WalletId -> Store
         (SqlPersistT IO)
-        (DeltaMap W.WalletId DeltaTxMetaHistory)
-mkStoreWalletsMeta =
+        (DeltaWalletsMetaWithSubmissions)
+mkStoreWalletMetaWithSubmissions wid =
     Store
     { loadS = load
     , writeS = write
     , updateS = update
     }
   where
-    write reset = forM_ (Map.assocs reset) $ \(wid,metas)
-        -> writeS (mkStoreMetaTransactions wid) metas
-    update _ (Insert wid metas) =
+    write (metas,subs) = do
         writeS (mkStoreMetaTransactions wid) metas
+        writeS (mkStoreSubmissions wid) subs
+    update (metas, TxLocalSubmissionHistory subs) xda = do
+        localTxToBeRemoved <- case xda of
+            ChangeMeta da -> do
+                updateS (mkStoreMetaTransactions wid) undefined da
+                let TxMetaHistory metas' = apply da metas
+                pure $ Set.difference (Map.keysSet subs) (Map.keysSet metas')
+            ChangeSubmissions da ->
+                updateS (mkStoreSubmissions wid) undefined da $> mempty
```

Comment:
```suggestion
    update (metas, subs) xda = do
        localTxToBeRemoved <- case xda of
            ChangeMeta da -> do
                updateS (mkStoreMetaTransactions wid) undefined da
                let TxLocalSubmissionHistory subs' = subs
                    TxMetaHistory metas' = apply da metas
                pure $ Set.difference (Map.keysSet subs') (Map.keysSet metas')
            ChangeSubmissions da ->
                updateS (mkStoreSubmissions wid) undefined da
                let TxLocalSubmissionHistory subs' = apply da subs
                    TxMetaHistory metas' = metas
                pure $ Set.difference (Map.keysSet subs') (Map.keysSet metas')
```

I have expanded the second case to match the behavior of `constraintSubmissions` exactly — we need to delete newly inserted local Tx submissions if there is no corresponding `TxMeta`. (It's true that this case would throw an error whenever the foreign key constraint is on. Also, we don't test this case. However, we should still handle it to be perfectly consistent with the implementation of `apply` for `ChangeSubmissions`.)

https://github.com/input-output-hk/cardano-wallet/blob/a3e76bc14dc302021c9aaa150eb4161b02117096/lib/core/src/Cardano/Wallet/DB/Store/Wallets/Model.hs#L117-L118